erpnext/stock/stock_ledger.py

```diff
@@ -453,20 +516,52 @@ def get_stock_ledger_entries(previous_sle, operator=None,
 			"order": order
 		}, previous_sle, as_dict=1, debug=debug)
 
-def get_valuation_rate(item_code, warehouse, voucher_type, voucher_no,
-	allow_zero_rate=False, currency=None, company=None, raise_error_if_no_rate=True):
+def get_valuation_rate(item_code, warehouse, voucher_type, voucher_no, batch_no=None,
+	allow_zero_rate=False, currency=None, company=None, raise_error_if_no_rate=True, batch_wise_valuation=None):
 	# Get valuation rate from last sle for the same item and warehouse
 	if not company:
 		company = erpnext.get_default_company()
 
-	last_valuation_rate = frappe.db.sql("""select valuation_rate
-		from `tabStock Ledger Entry`
-		where
-			item_code = %s
-			AND warehouse = %s
-			AND valuation_rate >= 0
-			AND NOT (voucher_no = %s AND voucher_type = %s)
-		order by posting_date desc, posting_time desc, name desc limit 1""", (item_code, warehouse, voucher_no, voucher_type))
+	last_valuation_rate = None
+
+	if batch_wise_valuation is None:
+		valuation_method, batch_wise_valuation = get_valuation_method(item_code)
+
+	if batch_no and batch_wise_valuation:
```

Comment:
move this to a smaller function

Reply:
Are you talking about this condition: `if batch_no and batch_wise_valuation` or the query?

Reply:
Alright